ocrd_network/ocrd_network/deployer.py

```diff
@@ -0,0 +1,310 @@
+
+
+from ocrd_utils import getLogger
+from ocrd_network.deployment_config import *
+from ocrd_network.deployment_utils import (
+    create_docker_client,
+    create_ssh_client,
+    CustomDockerClient,
+    DeployType,
+    HostData,
+    is_bashlib_processor,
+)
+from ocrd_network.rabbitmq_utils import RMQPublisher
+
+
+class Deployer:
+    """Wraps the deployment functionality of the Processing Server
+
+    Deployer is the one acting.
+    :py:attr:`config` is for representation of the config file only.
+    :py:attr:`hosts` is for managing processor information, not for actually processing.
+    """
+
+    def __init__(self, config: ProcessingServerConfig) -> None:
+        """
+        Args:
+            config (:py:class:`ProcessingServerConfig`): parsed configuration of the Processing Server
+        """
+        self.log = getLogger(__name__)
+        self.config = config
+        self.hosts = HostData.from_config(config.hosts)
+        self.mongo_pid = None
+        self.mq_pid = None
+
+    def kill_all(self) -> None:
+        """ kill all started services: workers, database, queue
+
+        The order of killing is important to optimize graceful shutdown in the future. If RabbitMQ
+        server is killed before killing Processing Workers, that may have bad outcome and leave
+        Processing Workers in an unpredictable state
+        """
+        self.kill_hosts()
+        self.kill_mongodb()
+        self.kill_rabbitmq()
+
+    def deploy_hosts(self, rabbitmq_url: str, mongodb_url: str) -> None:
+        for host in self.hosts:
+            self.log.debug(f'Deploying processing workers on host: {host.config.address}')
+
+            if (any(p.deploy_type == DeployType.native for p in host.config.processors)
+                    and not host.ssh_client):
+                host.ssh_client = create_ssh_client(
+                    host.config.address,
+                    host.config.username,
+                    host.config.password,
+                    host.config.keypath
+                )
+            if (any(p.deploy_type == DeployType.docker for p in host.config.processors)
+                    and not host.docker_client):
+                host.docker_client = create_docker_client(
+                    host.config.address,
+                    host.config.username,
+                    host.config.password,
+                    host.config.keypath
+                )
+
+            for processor in host.config.processors:
+                self._deploy_processing_worker(processor, host, rabbitmq_url, mongodb_url)
+
+            if host.ssh_client:
+                host.ssh_client.close()
+            if host.docker_client:
+                host.docker_client.close()
+
+    def _deploy_processing_worker(self, processor: WorkerConfig, host: HostData,
+                                  rabbitmq_url: str, mongodb_url: str) -> None:
+
+        self.log.debug(f'deploy \'{processor.deploy_type}\' processor: \'{processor}\' on'
```

Comment:
The main reason why all double quotes were replaced with single quotes is because that was requested earlier - to use single quotes for strings, hence, came the e16e3d42df04129864f1b3acf33868c70d9a9cb0 change in the past.

My default for strings is always double quotes for the exact same reason to not have to escape the single quotes in error messages. I like to avoid escaping characters due to bad readability.